main.go

```diff
@@ -29,63 +36,194 @@ func getType(body []byte) string {
 		log.Panic(err)
 	}
 
-	return service.Datacenter.Type
+	return service.Name, service.Datacenter.Type, service.Previous
 }
 
-func route(msg *nats.Msg, action string) []byte {
-	mapper := ""
-	switch getType(msg.Data) {
-	case "vcloud", "fake", "vcloud-fake":
-		mapper = "vcloud"
-	case "aws", "aws-fake":
-		mapper = "aws"
-	default:
-		return []byte(`{"error":"Invalid type"}`)
+func definitionToGraph(m libmapper.Mapper, body []byte) (*graph.Graph, error) {
+	var gd map[string]interface{}
+	err := json.Unmarshal(body, &gd)
+	if err != nil {
+		return nil, err
+	}
+
+	definition, ok := gd["service"].(map[string]interface{})
+	if ok != true {
+		return nil, errors.New("could not convert definition")
+	}
+
+	d, err := m.LoadDefinition(definition)
+	if err != nil {
+		return nil, err
 	}
 
-	subject := "definition.map." + action + "." + mapper
-	msg, err := n.Request(subject, msg.Data, time.Second)
+	return m.ConvertDefinition(d)
+}
+
+func mappingToGraph(m libmapper.Mapper, body []byte) (*graph.Graph, error) {
+	var gm map[string]interface{}
+	err = json.Unmarshal(body, &gm)
 	if err != nil {
-		log.Println("Error processing " + subject)
-		log.Println(err.Error())
+		return nil, err
 	}
 
-	return msg.Data
+	return m.LoadGraph(gm)
 }
 
 // SubscribeCreateService : definition.map.creation subscriber
-func SubscribeCreateService(msg *nats.Msg) {
-	res := route(msg, "creation")
-	if err := n.Publish(msg.Reply, res); err != nil {
-		log.Panic(err)
+// For a given definition, it will generate the valid service
+// and necessary workflow to create the environment on the
+// provider
+func SubscribeCreateService(body []byte) ([]byte, error) {
+	_, t, p := getInputDetails(body)
+
+	m := providers.NewMapper(t)
+	if m == nil {
+		return body, fmt.Errorf("Unconfigured provider type : '%s'", t)
```

Comment:
Should be logged by [this log.Panic(err.Error())](https://github.com/ernestio/definition-mapper/pull/12/files#diff-7ddfb3e035b42cd70649cc33393fe32cR188) :/